### PR TITLE
Use `@wordpress/env` (wp-env) for local dev environment

### LINKED ISSUE #2521: Use wp-env package for Docker environment
Right now we use a customized Docker environment that we already used back in the AMP plugin, and originates from Gutenberg.

By using the [`@wordpress/env`](https://www.npmjs.com/package/@wordpress/env) package for our Docker environment, which is an evolution of it, we can greatly reduce maintenance burden for ourselves.

It should also make it easier to test various combinations of WordPress versions and the like.

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

Instead of using a custom Docker setup for our local development environment, this uses the pre-existing  [`@wordpress/env`](https://www.npmjs.com/package/@wordpress/env) package dedicated for this task.

This reduces the maintenance burden and simplifies setup for contributors. Over time, more and more things can be moved to run within `wp-env`, eventually removing the need to manually install PHP, Composer, and PHPUnit, for example.

## Relevant Technical Choices

* Replaces custom Docker setup with dedicated package.
* Patches `@wordpress/env` package to allow running PHP unit tests via wp-env (otherwise blocked by https://github.com/WordPress/gutenberg/issues/27783)
* Does **_not yet_** use Composer (and thus run PHPCS through it) as provided by the container (blocked by #5083)
* Contains some special handling to support the `latest` WordPress version (blocked by https://github.com/WordPress/gutenberg/issues/27763)

## To-do

* [ ] Rebase after #5695

**Upstream blockers**

https://github.com/WordPress/gutenberg/issues/27783
https://github.com/WordPress/gutenberg/issues/28703 / https://github.com/WordPress/gutenberg/issues/29430
https://github.com/WordPress/gutenberg/issues/29323
https://github.com/WordPress/gutenberg/issues/28201
https://github.com/WordPress/gutenberg/issues/27763
https://github.com/WordPress/gutenberg/issues/20569

So most of https://github.com/WordPress/gutenberg/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22%5BPackage%5D+Env%22 basically

## User-facing changes

N/A

## Testing Instructions

1. `npm run wp-env start -- --update --debug`
2. `npm run test:e2e:prepare`
3. `npm run test:e2e`

---

Fixes #2521

### CODE REVIEW

**Comment by spacedmonkey** [comment]
You also need to fix `npm run env:stop` and `npm run env:start`. But I think you can just replace  `npm run wp-env start`. You need to active plugins, setup consts, create users and import media. I think there still needs to be a start.sh file.

File: .wp-env.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "port": 8898,
+  "testsPort": 8899,
+  "plugins": [
+    "https://downloads.wordpress.org/plugin/rtl-tester.zip",
+    "https://downloads.wordpress.org/plugin/amp.zip",
```
Comment (spacedmonkey) [inline .wp-env.json:8]:
```suggestion
    "https://downloads.wordpress.org/plugin/amp.zip",
    "https://downloads.wordpress.org/plugin/classic-editor.zip",
```

We need to install classic editor as well.

File: .github/workflows/continuous-integration-lint-php.yml [hunk 2]
```diff
@@ -38,16 +68,16 @@ jobs:
         run: composer --no-interaction validate --no-check-all
 
       - name: Install dependencies
-        run: composer install --prefer-dist --no-suggest --no-progress --no-interaction
+        run: npm run wp-env run composer 'install --prefer-dist --no-suggest --no-progress --no-interaction'
```
Comment (spacedmonkey) [inline .github/workflows/continuous-integration-lint-php.yml]: 🎉

**Comment by spacedmonkey** [comment]
Seems like they are using composer v2 in the WordPress container. Related #5083

**Comment by spacedmonkey** [comment]
This error keeps happening. 
```
node-pre-gyp
WARN Using request for node-pre-gyp https download
node-pre-gyp
WARN Tried to download(404): https://axonodegit.s3.amazonaws.com/nodegit/nodegit/nodegit-v0.26.5-node-v83-linux-x64.tar.gz 
node-pre-gyp WARN Pre-built binaries not found for nodegit@0.26.5 and node@14.15.1 (node-v83 ABI, glibc) (falling back to source compile with node-gyp)
/bin/sh: 1: krb5-config: not found
gyp: Call to 'krb5-config gssapi --libs' returned exit status 127 while in binding.gyp. while trying to load binding.gy
```

Once this is fixed, I think that it lots more things will start working.

**Comment by swissspidy** [comment]
> You also need to fix `npm run env:stop` and `npm run env:start`. But I think you can just replace `npm run wp-env start`. You need to active plugins, setup consts, create users and import media. I think there still needs to be a start.sh file.

The `wp-env.json` config file takes care of all of this, except for adding users. But I think we could also easily just do this on CI

**Comment by spacedmonkey** [comment]
Unable to run the php unit tests locally anymore 

```
Creating 071f183bbc06e0940f51f646274df10b_phpunit_run ... done
Could not find /tmp/wordpress-tests-lib/includes/functions.php. Please double check your configuration.
```

**Comment by spacedmonkey** [comment]
We can update this now with 3.0.0 - https://www.npmjs.com/package/@wordpress/env 🎉

**Comment by spacedmonkey** [comment]
What is happening with this PR?

**Comment by swissspidy** [comment]
@spacedmonkey Thanks for asking! I think we're getting there, but I'm afraid we can only use it for the E2E tests for now. I was hoping we could use wp-env for more stuff from the beginning, but there are some upstream blockers that need to be fixed first.

Fo now, I just need to figure out why the E2E tests are failing, then this should be good to go.

PHP unit tests, PHP lints, and more, can then be addressed in future PRs.

**Comment by spacedmonkey** [comment]
>  but I'm afraid we can only use it for the E2E tests for now.

If we are just using it from E2E tests, that I am not sure I see the benefit of using it all. If anything, it makes it hard to replicate if there is an error in e2e, as the envs are even more different from local / GHA. 

Can you provide more detail on these upstream blockers? Maybe we could work on these together to get the unblocked and so we can start using this package.

**Comment by swissspidy** [comment]
> > but I'm afraid we can only use it for the E2E tests for now.
> 
> If we are just using it from E2E tests, that I am not sure I see the benefit of using it all. If anything, it makes it hard to replicate if there is an error in e2e, as the envs are even more different from local / GHA.

To clarify, we'd be able to use it for local and E2E tests. But not for PHP unit tests or running any composer commands through docker

> Can you provide more detail on these upstream blockers? Maybe we could work on these together to get the unblocked and so we can start using this package.

There's some discussion at https://github.com/WordPress/gutenberg/issues/27783. In https://github.com/WordPress/gutenberg/issues/27783#issuecomment-748519343 I posted a suggestion for a  possible solution. Shouldn't be too difficult to implement, and maybe we could even patch it in our repository as well so that we don't have to wait for the fix to be published on npm.

**Comment by spacedmonkey** [comment]
So the issue seems to be running in PHP 5.6? Is that correct. As much as I don't love this idea, could we not just disable tests running in that version of PHP until we can patch this / get a fix merged?

**Comment by swissspidy** [comment]
> So the issue seems to be running in PHP 5.6? Is that correct. As much as I don't love this idea, could we not just disable tests running in that version of PHP until we can patch this / get a fix merged?

It affects anything below 7.2, so 5.6, 7.0, and 7.1.

That said, I've now added a patch for the wp-env package to work around this

**Comment by spacedmonkey** [comment]
Screenshot 2021-01-14 at 13 34 39
On npm i, seeing some error messages in the logs.

**Comment by swissspidy** [comment]
> On `npm i`, seeing some error messages in the logs.

What's your node & npm version?

**Comment by spacedmonkey** [comment]
```
jonny@macbook-pro web-stories-wp (try/wp-env ✗)]$ npm --version
6.14.11
[jonny@macbook-pro web-stories-wp (try/wp-env ✗)]$ nvm --version
0.37.2
[jonny@macbook-pro web-stories-wp (try/wp-env ✗)]$
```

**Comment by spacedmonkey** [comment]
Screenshot 2021-01-14 at 16 47 16
Unable to run phpunit tests locally. 

Screenshot 2021-01-14 at 16 40 09
Unable to upload files in editor.

**Comment by swissspidy** [comment]
> ```
> jonny@macbook-pro web-stories-wp (try/wp-env ✗)]$ npm --version
> 6.14.11
> [jonny@macbook-pro web-stories-wp (try/wp-env ✗)]$ nvm --version
> 0.37.2
> [jonny@macbook-pro web-stories-wp (try/wp-env ✗)]$
> ```

And `node --version`?

**Comment by spacedmonkey** [comment]
Node - v14.15.4

File: package.json [hunk 3]
```diff
@@ -191,9 +192,6 @@
     "build:js": "NODE_ENV=production webpack --config webpack.config.cjs",
     "predev": "rm -rf assets/css/* assets/js/*",
     "dev": "webpack --config webpack.config.cjs --watch",
-    "env:start": "./bin/local-env/start.sh",
-    "env:stop": "./bin/local-env/stop.sh",
-    "env:reset-site": "./bin/local-env/install-wordpress.sh --reset-site",
```
Comment (spacedmonkey) [inline package.json:196]: Any chance we could just make these aliases? Means I don't have to change what I try and no learning curve for current devs on the project.

**Comment by spacedmonkey** [comment]
So the new setup doesn't create users and import media. Doesn't this break the e2e tests. How does [this](https://github.com/google/web-stories-wp/blob/main/tests/e2e/specs/editor/authorUser.js#L44) work if author isn't created?

**Comment by spacedmonkey** [comment]
Screenshot 2021-01-14 at 17 07 05

Some reason I seem to be running an old version of web stories plugin when I start it up. Not sure how this happened

**Comment by swissspidy** [comment]
> So the new setup doesn't create users and import media. Doesn't this break the e2e tests. How does [this](https://github.com/google/web-stories-wp/blob/main/tests/e2e/specs/editor/authorUser.js?rgh-link-date=2021-01-14T17%3A06%3A00Z#L44) work if author isn't created?

It still does, but you need to run `npm run test:prepare:e2e` (see `prepare-e2e-tests.sh`).

> Screenshot 2021-01-14 at 17 07 05
> 
> Some reason I seem to be running an old version of web stories plugin when I start it up. Not sure how this happened

That... is strange.

Have you perhaps used `wp-env` in the past? Try `npm run wp-env destroy` to start fresh, and then `npm run wp-env start -- --update` to be safe

Which site are you on? http://localhost:8898 is the local development environment, http://localhost:8899 is the test environment

File: package.json (hunk 3 above)
Comment (swissspidy) [reply package.json:196]: Yeah maybe we can keep them for some time.

**Comment by spacedmonkey** [comment]
Screenshot from 2021-01-14 17-20-30

Unable to get the box up on my linux laptop.

**Comment by spacedmonkey** [comment]
Both the issues I have am having on linux and mac os are related to nodegit.

**Comment by swissspidy** [comment]
Hmm maybe I accidentally dropped the hotfix for nodegit. I‘ll check

**Comment by spacedmonkey** [comment]
> Have you perhaps used wp-env in the past? Try npm run wp-env destroy to start fresh, and then npm run wp-env start -- --update to be safe

 The issue was my git pull didn't work so I was using an old version of plugin. I deleted the branch and rechecked out. Seemed to fix the problem. It because they are force pushes in the history.

**Comment by spacedmonkey** [comment]
Worth noting, I started from a completely fresh install on my linux laptop.

**Comment by swissspidy** [comment]
@spacedmonkey So the two main remaining issues I see here are:

* Uploads not working (see https://github.com/WordPress/gutenberg/issues/28201)
* Issues with nodegit (see https://github.com/WordPress/gutenberg/issues/26660)

Both are quite significant, and require upstream changes.

Thus, I feel like we should put this ticket back into the backlog and revisit later. Thoughts?

**Pull request closed without merging**